### ISSUE TITLE
Question on Ghost Batch Norm

### ISSUE THREAD

**Issue by getgaurav2** [dreamquark-ai/tabnet#384]
Hello @Optimox  - 

Quick question on the Ghost BatchNorm  : In case x.shape[0] has values like ( n * virtual_batch_size + 1 )  then the last value shall be left to fed into the Batch Norm on its own and it will throw an error. 

For Example : 
x.shape[0] = 101
virtual_batch_size = 10 

This will create 11 chunks for x , where the 11th chunk will have only 1 vector and will throw an error when passed into `self.bn(x_)` 

"""

       class GBN(torch.nn.Module):
      
          Ghost Batch Normalization
          https://arxiv.org/abs/1705.08741
     
  
          def __init__(self, input_dim, virtual_batch_size=128, momentum=0.01):
              super(GBN, self).__init__()
  
              self.input_dim = input_dim
              self.virtual_batch_size = virtual_batch_size
              self.bn = BatchNorm1d(self.input_dim, momentum=momentum)
  
          def forward(self, x):
              chunks = x.chunk(int(np.ceil(x.shape[0] / self.virtual_batch_size)), 0)
              res = [self.bn(x_) for x_ in chunks]
  
              return torch.cat(res, dim=0)

"""

Question : Can the line below be a suitable replacement ?

`res = [ self.bn(x_)  if x_.shape[0]> 1 else  x_ for x_ in chunks]`

**Comment by Optimox** [comment]
Hello @getgaurav2,

Indeed if the training size modulo the batch size is 1 you can't compute the batch normalization.

The easiest workaround is to set `drop_last=True` to be sure to avoid this.

Your workaround would prevent the error, but won't normalize the last sample which would be a problem as the inputs are not normalized so the batch normalization can make a significant change to the inputs.

I guess it would be better to do something like this (dropping last samples when the rest of division is 1):
`res = [ self.bn(x_)  for x_ in chunks  if x_.shape[0]> 1]`

However, for this to work we would need to do the same with the targets. It would not be a very elegant solution code wise.
Another solution would be to duplicate the last sample, apply batch normalization and then keep only one example, but it's not very elegant either.

I think the best is to set `drop_last=True`, it will be the case by default in the next release.
